_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,295 @@
+            }
+          }
+        },
+        {
+          "neural": {
+            "title_embedding": {
+              "query_text": "science fiction",
+              "model_id": "aVeif4oB5Vm0Tdw8zYO2",
+              "k": 5
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+This query uses:
+
+- `hybrid`: Combines results from lexical and semantic search.
+- `search_pipeline`: Applies the RRF combination logic to rank the final results.
+
+### How RRF works
+
+RRF works as follows:
+
+1. **Sort documents by score**: Each query method sorts documents by score.
+1. **Assign rank positions**: Documents are ranked based on their scores for each query.
+1. **Compute the RRF score**: For each document, the RRF score is computed using the following formula:
+    ```python
+    rankScore(document_i) = sum((1/(k + query_1_rank), (1/(k + query_2_rank), ..., (1/(k + query_j_rank)))
+    ```
+
+    where:
+    -  `k` is a rank constant 
+    - `query_j_rank` represents the ranking of a document in a particular query method. 
+
+1. **Add rank contributions**: Rank calculations are combined, and documents are sorted by decreasing rank score.
+1. **Return the top results**: The highest-ranked documents are retrieved based on the query size.
+
+## When to use rank-based normalization
+
+Consider using rank-based normalization in the following situations:
+
+- **Heterogeneous score distributions**: Rank fusion (especially RRF) excels when BM25 and semantic models produce scores on incompatible scales or with outliers. It avoids complex score calibration by using result ranks only. This yields *stable rankings* even if one query method’s scores would otherwise dominate after normalization.
+- **No tuning or calibration needed**: RRF is an out-of-the-box method: it requires no pre-training, weight tuning, or knowledge of score ranges. In general-purpose search systems (with diverse queries and content), it's often impractical to hand-tune weighting for every scenario. Rank fusion is a robust default when labeled data for calibration is unavailable.
+* **Resilient to outliers and domain shifts**: Because RRF ignores absolute score magnitude, an extreme-scoring outlier won't skew the results. This is useful in general collections, in which some queries or documents might produce anomalously high scores using one model. Similarly, if the data distribution or query mix changes over time, RRF remains stable without needing recalibration because it uses relative rank positions.
+* **Adaptable to changing environments:** When the nature of queries and/or data evolves over time, score-based methods often require continuous fine-tuning of weights to maintain high relevancy. RRF eliminates this need for ongoing maintenance because it relies on relative rankings rather than absolute scores. This makes it particularly valuable in dynamic environments where constant recalibration is resource-intensive or impractical.
```

Comment:
```suggestion
* **Adaptable to changing environments:** When the nature of queries and/or data evolves over time, score-based methods often require continuous fine-tuning of weights to maintain high relevancy. RRF eliminates this need for ongoing maintenance because it relies on relative rankings rather than absolute scores. This makes it particularly valuable in dynamic environments where constant recalibration is resource intensive or impractical.
```